### PR TITLE
Allow setting timeouts for NFS operations

### PR DESCRIPTION
**Pull request by kibab**
Since NFS server may disappear or start being unresponsive at any time,
make it possible to set a TCP timeout when making requests.
After getting timeout error, the connection most likely has to be restarted.

To make this change possible, specify the "real" type of "wc" as net.Conn
instead of hiding all useful functionality behind the interface specification.

### CODE REVIEW

File: nfs/rpc/tcp.go [hunk 1]
```diff
@@ -51,3 +62,11 @@ func (t *tcpTransport) Write(buf []byte) (int, error) {
 func (t *tcpTransport) Close() error {
 	return t.wc.Close()
 }
+
+func (t *tcpTransport) SetTimeout(d time.Duration) {
+	t.timeout = d
+	if d == 0 {
+		var zeroTime time.Time
+		t.wc.SetDeadline(zeroTime)
```
Comment (fdawg4l) [inline nfs/rpc/tcp.go:70]: Just curious, what is the expected behavior here? If set to 0, will the connection timeout immediately?

**Review by fdawg4l — APPROVED** [review]

File: nfs/rpc/tcp.go (hunk 1 above)
Comment (kibab) [reply nfs/rpc/tcp.go:70]: No, zero means "no timeout at all", this is a documented behavior of SetDeadline(), see https://golang.org/pkg/net/, a comment above SetDeadline()

Comment (fdawg4l) [reply nfs/rpc/tcp.go:70]: Great! Thanks!

Comment (matthewavery) [reply nfs/rpc/tcp.go:70]: Hey @fdawg4l :) I believe the behavior if set to `0` is that the connection will never timeout. This is dictated by the behavior of the `Conn` struct [here](https://golang.org/pkg/net/#Conn).

Comment (matthewavery) [reply nfs/rpc/tcp.go:70]: ah I should have refreshed the page.

**Review by matthewavery — APPROVED** [review]

File: nfs/rpc/tcp.go (hunk 1 above)
Comment (kibab) [reply nfs/rpc/tcp.go:70]: @matthewavery Yeah, that's why I'm adding SetTimeout() calls in recv()/Write() methods -- so the caller just needs to call SetTimeout() once and then the code will update the deadlines and Do The Right Thing (c)

**Comment by matthewavery** [comment]
@kibab thanks for the contribution! Happy Hacktoberfest!

File: nfs/rpc/tcp.go (hunk 1 above)
Comment (matthewavery) [reply nfs/rpc/tcp.go:70]: yes, I missed the write side of things in my first glance through. My apologies.

**Pull request merged**